### PR TITLE
NPT support

### PR DESCRIPTION
**Pull request by maxentile**
Closes https://github.com/proteneer/timemachine/issues/373 by cloning MonteCarloBarostat implementation from OpenMM.

- [x] Add [`NVTEnsemble`](https://github.com/proteneer/timemachine/blob/3e20a5a18d4b8aaeba144001fd044a08b03417cb/md/ensembles.py#L52-L78) and [`NPTEnsemble` ](https://github.com/proteneer/timemachine/blob/3e20a5a18d4b8aaeba144001fd044a08b03417cb/md/ensembles.py#L81-L121) definitions.
- [x] Add [`MonteCarloMove`](https://github.com/proteneer/timemachine/blob/3e20a5a18d4b8aaeba144001fd044a08b03417cb/md/moves.py#L6-L31)s for temperature control ([`UnadjustedLangevinMove`](https://github.com/proteneer/timemachine/blob/3e20a5a18d4b8aaeba144001fd044a08b03417cb/md/thermostat/moves.py#L7-L34)) and pressure control ([`MonteCarloBarostat`](https://github.com/proteneer/timemachine/blob/3e20a5a18d4b8aaeba144001fd044a08b03417cb/md/barostat/moves.py#L86-L196)). NPT simulation is performed by [composing](https://github.com/proteneer/timemachine/blob/3e20a5a18d4b8aaeba144001fd044a08b03417cb/md/utils.py#L25) these moves. Later extensions of timemachine might update the current UnadjustedLangevinMove to use the LangevinMiddleIntegrator scheme from OpenMM, or add wrappers for external implementations of more advanced HMC moves.
- [x] Add [usage example](https://github.com/proteneer/timemachine/blob/3e20a5a18d4b8aaeba144001fd044a08b03417cb/examples/npt_equilibration.py) applying NPT equilibration to each window of an absolute free energy calculation.
- [x] Add [tests](https://github.com/proteneer/timemachine/blob/3e20a5a18d4b8aaeba144001fd044a08b03417cb/tests/test_ensembles.py) of the reduced potential computation in NVT, NPT ensembles
- [x] Add [test](https://github.com/proteneer/timemachine/blob/3e20a5a18d4b8aaeba144001fd044a08b03417cb/tests/test_barostat.py#L56-L155) that barostat'd molecular ideal gas volume is consistent with expectation
- [x] Add [slow test](https://github.com/proteneer/timemachine/blob/3e20a5a18d4b8aaeba144001fd044a08b03417cb/slow_tests/test_barostat_slow.py) that barostat'd density of water is very close to 1.0 kg/L
- [x] Add [convenience function](https://github.com/proteneer/timemachine/blob/3e20a5a18d4b8aaeba144001fd044a08b03417cb/md/thermostat/utils.py#L6-L19) for sampling according to temperature, and delete vestigial [`set_velocities_to_temperature`](https://github.com/proteneer/timemachine/commit/1f976288230b5c429cf0568387d855f873bf6c70#diff-c8d76629b43f57a44bd424b60d79e55e4bf555c7031210b0f15bf35783071bd7L8-L13)
- [x] Delete vestigial [barostat/barostat.py](https://github.com/proteneer/timemachine/blob/0ff865c8c84d43c8a76e2a1ff0e3c178dc5fb7ad/barostat/barostat.py), [barostat/potentials.py](https://github.com/proteneer/timemachine/blob/0ff865c8c84d43c8a76e2a1ff0e3c178dc5fb7ad/barostat/potentials.py)

### CODE REVIEW

File: slow_tests/test_barostat_slow.py
Comment (badisa) [inline slow_tests/test_barostat_slow.py:27]: Wrapping this in a unittest.TestCase seems cleaner than having this as a script

**Review by badisa — APPROVED** [review]

File: slow_tests/test_barostat_slow.py
Comment (badisa) [reply slow_tests/test_barostat_slow.py:27]: Actually trying to run this, it took ~5 minute so perhaps best not as a test? Might be worth deciding on some convention between a 'test' and a 'benchmark'. As the slow tests still get run in CI

Comment (maxentile) [reply slow_tests/test_barostat_slow.py:27]: I had mistakenly thought CI was running  `slow_tests/` less frequently than `tests/` (but I see that the actual behavior of `slow_tests/` is documented in [this PR description](https://github.com/proteneer/timemachine/pull/405) -- tests in this folder are run without cuda-memcheck so that they can be run at the same frequency as `tests/`, just faster). Happy to move this wherever is appropriate.

Do you think we should move this from `slow_tests/` into a new `benchmarks/` folder or something that doesn't get discovered by CI at all?

Another alternative might be to [mark](https://docs.pytest.org/en/6.2.x/mark.html) this and possibly other tests as slow, telling pytest to skip these most of the time, but periodically still run the slow tests through pytest.

Comment (badisa) [reply slow_tests/test_barostat_slow.py:27]: I say we leave this for now, as it is, this won't cause any issues. Probably warrants more discussion than a PR would allow

File: examples/npt_equilibration.py
Comment (proteneer) [inline examples/npt_equilibration.py:41]: Should we be consistent with OpenMM for sanity purposes? IIRC the default was 1 bar (which is very close but not exactly to 1 atm)

File: md/barostat/moves.py
Comment (proteneer) [inline md/barostat/moves.py:74]: Does the OpenMM centroid computation take mass into account? @badisa

Comment (proteneer) [reply md/barostat/moves.py:74]: (PS if we decide to support this, `np.average` as opposed to `np.mean` allows one to pass in a list of weights.

File: md/barostat/moves.py
Comment (proteneer) [inline md/barostat/moves.py:160]: I like these tweaks!

File: md/barostat/moves.py
Comment (badisa) [reply md/barostat/moves.py:74]: Based on my reading of the OpenMM, it does not take the masses into account. That said, the kernel logic that they have is  fairly opaque to me. 

https://github.com/openmm/openmm/blob/be19e0222ddf66f612016a3c1f687161a53c2396/platforms/common/src/CommonKernels.cpp#L7625

File: md/barostat/utils.py
Comment (proteneer) [inline md/barostat/utils.py:10]: Nice! I hadn't realized `det(box)` was the proper way to generalize volumes to triclinic systems, but after some searching it appears you are correct.

https://journals.iucr.org/a/issues/2011/01/00/au5114/au5114.pdf

File: md/barostat/utils.py
Comment (proteneer) [inline md/barostat/utils.py:36]: is this still being used?

File: md/barostat/utils.py
Comment (proteneer) [inline md/barostat/utils.py:25]: Note that in certain cases (eg. core restraints for ABFE), we will actually have two harmonic bond terms, one for the regular valences, and one for the restraining term. It's probably more flexible to just modify the input of this function to either

1) a list of HarmonicBonds
or perhaps slightly cleaner:
2) have it take in a simple Bx2 list of idxs (and leave it up to the caller to pass merge the indices).

File: md/thermostat/utils.py
Comment (proteneer) [inline md/thermostat/utils.py:14]: I double checked this against https://github.com/choderalab/openmm-validation/blob/master/tests/test_custom_integrators.py#L62

and it seems consistent! nice.

**Comment by proteneer** [comment]
One last thing I would add (and probably something you two are already thinking of) - can you work with @badisa on figuring what unit tests for his C++ code should be for correctness purposes?

**Review by proteneer — APPROVED** [review]
LGTM - some clerical comments but otherwise looks fantastic. This also sets up a good framework in case we want to introduce other types of MC moves in the future! (eg. random rigid rotation move, etc.)

File: examples/npt_equilibration.py
Comment (maxentile) [reply examples/npt_equilibration.py]: Updated the examples from atm to bar in https://github.com/proteneer/timemachine/pull/433/commits/0f483831b931ee8ed3844391a9aa5add835290ca. (Note that we don't set a default temperature or pressure in the `barostat` or `ensembles` code itself, and allow user inputs in any simtk pressure unit.)

File: md/barostat/utils.py
Comment (maxentile) [reply md/barostat/utils.py]: Ahh, good catch -- this had been used in an earlier version where I wanted to have the barostat displace the protein:ligand complex as a single unit (rather than displacing the protein and ligand separately). Examples don't currently exercise this option, and I don't currently see a compelling reason to do it one way or the other, so I removed this function in https://github.com/proteneer/timemachine/pull/433/commits/03952ce62b88fecf1a7c43408413db7ea9796175 .

File: md/barostat/utils.py
Comment (maxentile) [reply md/barostat/utils.py]: Updated the signature as suggested here https://github.com/proteneer/timemachine/pull/433/commits/954937ce6ba2a2b4c44a2c6d36c5efc59c9318e7 . Later, might be worthwhile to remove both of these and replace with an "alchemical topology" object or similar, that helps keep track of what bonds are present regardless of whether they're represented as harmonic bond terms, distance constraints, etc. down the road.

File: tests/test_barostat.py
Comment (badisa) [inline tests/test_barostat.py:138]: If you disable the barostat in this test, this test still passes

Comment (maxentile) [reply tests/test_barostat.py:139]: Explainable (but undesirable!) behavior, since the test is initialized at the expected volume here https://github.com/proteneer/timemachine/pull/433/commits/5f9cf6dccd2021060ec6033c260cc346dc0adc71 .

The initial version of the test took a long time to run, since it required to equilibrate to a far larger box size than initialized (like, initialized at ~10-100 nm^3, with ~10000 nm^3 expected). However, changing the initialization to be _exactly_ the expected box volume has the undesirable effect of making the test only check whether running the barostat causes the simulation to _drift away from the correct answer._ We of course also would want this test to fail if the barostat is doing nothing!

The behavior of the test should be fixed in https://github.com/proteneer/timemachine/pull/433/commits/05817c851e676ccb40042ca2b121adbfa367213b , where the barostat is initialized from a box that's instead 10% too large.

Comment (maxentile) [reply tests/test_barostat.py:139]: And with this aspect of the test behavior fixed, it appears the test is no longer passing!

This needs to be resolved before merging.

Possibilities:
* there are uncaught bugs with the barostat itself
* the test is not behaving as intended for another reason, e.g. because simulation with such a giant box is causing other numerical issues or slow thermostat convergence. (Note that the OpenMM version of this test system has only 64 particles, and much smaller expected box sizes...)

Comment (maxentile) [reply tests/test_barostat.py:139]: [Increasing the pressure by a factor of 100](https://github.com/proteneer/timemachine/pull/433/commits/4886d6ec176d26d70a8bdc2ee70aa9e951e89781) so that the expected volume is within a couple orders of magnitude of water box simulation cells we typically consider, the test is failing but with small relative differences from expectation.

```
Mismatched elements: 1 / 3 (33.3%)
Max absolute difference: 4.03597219
Max relative difference: 0.01881098
 x: array([107.001069, 218.590044, 360.868921])
 y: array([107.277036, 214.554072, 357.59012 ])
```

Will adjust to run this test longer...

Comment (maxentile) [reply tests/test_barostat.py:139]: Adjusting the molecular ideal gas tests to [run 5x longer](https://github.com/proteneer/timemachine/pull/433/commits/ba0e6676d8198b23244425e0c52a73963e207938) causes this to pass, reducing my level of suspicion of an uncaught bug in the barostat itself. However, this adds another 3 minutes to the tests...

Comment (jchodera) [reply tests/test_barostat.py:139]: Could limiting `max_delta_volume=3.0` here be a source of slow convergence? An optimal choice (or adaptively identified choice) would presumably converge pretty quickly if you're just testing the mean.

You can also cut your convergence time in half by not discarding the first half to equilibration, but instead starting at the analytical volume and seeing if it drifts away?

**Review by badisa — APPROVED** [review]
Looks good. Only thought would be to move the barostat test to the slow tests to avoid running under cuda memcheck as you had mentioned previously.

File: tests/test_barostat.py
Comment (maxentile) [reply tests/test_barostat.py]: Thank you! These comments are partially addressed in https://github.com/proteneer/timemachine/pull/433/commits/64520756b9f155edef466049d3f1f75e087895bd by removing the `max_delta_volume` argument, and initializing in a way that should require less equilibration time.

A previous version of this test did initialize at the analytical volume and wait to see if it drifts away, but this led to false passes. Here I've opted to initialize just slightly outside of the relative tolerance, so that the required convergence time is reduced, but the test will still fail if the barostat does nothing.

**Comment by maxentile** [comment]
>move the barostat test to the slow tests to avoid running under cuda memcheck as you had mentioned previously.

Thank you! Addressed in https://github.com/proteneer/timemachine/pull/433/commits/fef7fe1a01c913cd6fb13952b7eee093c2219d1e

**Comment by badisa** [comment]
Running the molecular ideal gas test on my dev box I see a failure:

```
E       AssertionError: 
E       Not equal to tolerance rtol=0.01, atol=0
E       
E       Mismatched elements: 1 / 3 (33.3%)
E       Max absolute difference: 2.78290284
E       Max relative difference: 0.0107958
E        x: array([107.813074, 216.870393, 360.373084])
E        y: array([107.277054, 214.554109, 357.590181])
```

**Comment by maxentile** [comment]
>Running the molecular ideal gas test on my dev box I see a failure:

Ahh, off by 1.07% where the tolerance was 1%... Reverting from `n_moves=2000` to `n_moves=10000` in https://github.com/proteneer/timemachine/pull/433/commits/c234ff4b60cac00a86a1415205714779e4ef13d3 fixes this for now, and later we can pick the relative tolerance in a bit more of a principled way based on the expected variability.

**Pull request merged**
